### PR TITLE
change how labels work; introduce qualifiers

### PR DESCRIPTION
**Pull request by krancour**
Potential fix for #1323 

cc @itowlson

If possible, I want to get this into v2-alpha.2, even if it means delaying the release a few days.

Note: Name "qualifiers" subject to change if anyone has a better suggestion.

### CODE REVIEW

File: v2/apiserver/internal/core/mongodb/projects_store.go
Comment (vdice) [inline v2/apiserver/internal/core/mongodb/projects_store.go:124]: Wow, accruing advanced mongo knowledge.  Hadn't seen this method.

File: v2/apiserver/internal/core/mongodb/projects_store.go
Comment (vdice) [inline v2/apiserver/internal/core/mongodb/projects_store.go:173]: I wonder if there's a way to cover under test?  (Could be in a follow-up).  Maybe we try to test the JS directly (not via mongo or e2e) with pre-loaded data, to verify its logic is correct/intended?

File: v2/apiserver/internal/core/events.go
Comment (vdice) [inline v2/apiserver/internal/core/events.go:78]: I notice how the `Qualifiers` field here has its type as `Qualifiers` (to utilize its `MarshalBSONValue` method, right?)  Curious as to why we don't have a need for the same for Labels -- it looks like it _used_ to have that same method, etc.?

File: v2/apiserver/internal/core/mongodb/projects_store.go
Comment (krancour) [reply v2/apiserver/internal/core/mongodb/projects_store.go:173]: Good question. That question actually applies to all our queries (i.e. not just this one on account of the JS). Our unit tests for store code all simulate responses from the database, but they don't account at all for the possibility that a query is simply _wrong_.

What I think I'm banking on here is that once Brigade 2 reaches a point where we're using it to test itself, the new support for jobs with sidecars means that we can stand up a live MongoDB instance alongside our Go test container and actually start to put the store queries to the test.

File: v2/apiserver/internal/core/events.go
Comment (krancour) [reply v2/apiserver/internal/core/events.go:78]: Good observation. It has to do with how each of the fields is queried.

Technically, the new `Qualifiers` field _is_ the old `Labels` field, simply renamed. The _new_ `Labels` field is the _real_ new thing.

When we match on qualifiers, we are looking for an _exact_ match. (This is what we _used_ to be doing with the labels, but @itowlson pointed out, quite correctly, that such behavior really didn't comport with reasonable assumptions people are likely to have about how "labels" work.) In MongoDB, an _exact_ match on a subdocument requires you to be using `bson.D` instead of `bson.M` and that's basically the reason for the `MarshalBSONValue` function on the `Qualifiers` (nee `Labels`) type. (The difference between `bson.D` and `bson.M` is that the order of the keys is canonicalized in a `bson.D`. `bosn.M` doesn't have that feature. Think of it like canonical JSON vs "regular" JSON.)

The match semantics on the _new_ `Labels` field don't involve making an exact match on a subdocment, so this concern of having to use `bson.D` doesn't present itself.

File: v2/apiserver/internal/core/mongodb/projects_store.go
Comment (krancour) [reply v2/apiserver/internal/core/mongodb/projects_store.go:124]: I just learned this trick in the last few days. I'd rather not have done this, but it's the only way I could find to do the "reverse" label match.

File: v2/apiserver/internal/core/events.go
Comment (vdice) [reply v2/apiserver/internal/core/events.go:78]: Ah, thank you.  That clarifies things!

**Review by vdice — APPROVED** [review]

**Comment by vdice** [comment]
Just had a thought for a follow-up:  It may be useful to add an example showing a project with an eventSubscription config utilizing the new qualifiers and labels fields (and maybe others?)

**Comment by krancour** [comment]
And fwiw, I still want @itowlson to weigh in on this before we merge.

**Review by itowlson — APPROVED** [review]
This is super useful - thank you.  I agree the naming might need to evolve but I don't have any ideas I'm afraid; not qualified to review the code without a longish ramp-up; so LGTM!

**Pull request merged**
